File: lib_appcommons/src/main/java/ng/kingsley/android/preference/PreferenceStore.java
package ng.kingsley.android.preference;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;

import ng.kingsley.android.util.DigestUtils;

/**
 * @author ADIO Kingsley O.
 * @since 14 Oct, 2016
 */

public class PreferenceStore {

    private SharedPreferences pref;
    private Gson gson;

    PreferenceStore(Context context, Gson gson, String domain) {
        this.gson = gson;
        pref = context.getSharedPreferences(name(context, domain), Context.MODE_PRIVATE);
    }

    PreferenceStore(SharedPreferences wrapped, Gson gson) {
        this.gson = gson;
        this.pref = wrapped;
    }

    private static String name(Context context, String domain) {
        domain = (domain == null) ? "" : domain.trim();
        return DigestUtils.md5(context.getPackageName() + domain);
    }

    private SharedPreferences.Editor editor() {
        return pref.edit();
    }

    public void putInt(String key, int value) {
        editor().putInt(key, value).apply();
    }

    public void putLong(String key, long value) {
        editor().putLong(key, value).apply();
    }

    public void putFloat(String key, float value) {
        editor().putFloat(key, value).apply();
    }

    public void putBoolean(String key, boolean value) {
        editor().putBoolean(key, value).apply();
    }

    public void putString(String key, String content) {
        editor().putString(key, content).apply();
    }

    public <T> void put(String key, T content, Converter<T> converter) {
        putString(key, converter.toString(content));
    }

    public <T> void put(String key, T content, Class<T> clas) {
        put(key, content, new DefaultConverters.GsonConverter<>(clas, gson));
    }

    public <T> void put(String key, T content, TypeToken<T> token) {
        put(key, content, new DefaultConverters.GsonConverter<>(token, gson));
    }

    public int getInt(String key, int defaultValue) {
        return pref.getInt(key, defaultValue);
    }

    public long getLong(String key, long defaultValue) {
        return pref.getLong(key, defaultValue);
    }

    public float getFloat(String key, float defaultValue) {
        return pref.getFloat(key, defaultValue);
    }

    public boolean getBoolean(String key, boolean defaultValue) {
        return pref.getBoolean(key, defaultValue);
    }

    public String getString(String key, String defaultValue) {
        return pref.getString(key, defaultValue);
    }

    public <T> T get(String key, Converter<T> converter) {
        return converter.fromString(getString(key, null));
    }

    public <T> T get(String key, Class<T> clas) {
        return get(key, new DefaultConverters.GsonConverter<>(clas, gson));
    }

    public <T> T get(String key, TypeToken<T> token) {
        return get(key, new DefaultConverters.GsonConverter<>(token, gson));
    }

    public void remove(String key) {
        editor().remove(key).apply();
    }

    public void clear() throws IOException {
        editor().clear().apply();
    }
}

File: preference/src/main/java/com/kingsleyadio/appcommons/preference/PreferenceStore.kt
package com.kingsleyadio.appcommons.preference

import android.content.SharedPreferences
import java.lang.reflect.Type

/**
 * @author ADIO Kingsley O.
 * @since 14 Oct, 2016
 */
class PreferenceStore internal constructor(
    private val pref: SharedPreferences,
    private val converter: Converter
) {

    fun getInt(key: String, defaultValue: Int): Int = pref.getInt(key, defaultValue)

    fun getLong(key: String, defaultValue: Long): Long = pref.getLong(key, defaultValue)

    fun getFloat(key: String, defaultValue: Float): Float = pref.getFloat(key, defaultValue)

    fun getBoolean(key: String, defaultValue: Boolean): Boolean = pref.getBoolean(key, defaultValue)

    fun getString(key: String): String? = pref.getString(key, null)

    fun getString(key: String, defaultValue: String): String = getString(key) ?: defaultValue

    fun <T> get(key: String, type: Type): T? = getString(key)?.let { converter.fromString(it, type) }

    fun putInt(key: String, value: Int): Unit = edit { putInt(key, value) }

    fun putLong(key: String, value: Long): Unit = edit { putLong(key, value) }

    fun putFloat(key: String, value: Float): Unit = edit { putFloat(key, value) }

    fun putBoolean(key: String, value: Boolean): Unit = edit { putBoolean(key, value) }

    fun putString(key: String, content: String?): Unit = edit { putString(key, content) }

    fun <T : Any> put(key: String, content: T, type: Type): Unit = putString(key, converter.toString(content, type))

    fun containsKey(key: String): Boolean = pref.contains(key)

    fun remove(key: String): Unit = edit { remove(key) }

    fun clear(): Unit = edit { clear() }

    private fun edit(action: SharedPreferences.Editor.() -> Unit) = pref.edit().apply(action).apply()
}
